### PR TITLE
Add static quantization as an example for calibration flow

### PR DESCRIPTION
**Pull request by jerryzh168**
Summary:
So far quantization flow API that we provided (`quantize_`) does not require calibration (calibrate a model with sample data), this PR added a static quantization example that serves as an example for calibration flow

* 1. first prepare the model for calibration
* 2. calibrate the prepared model with sample data
* 3. convert the calibrated model to quantized model

Test Plan:
python torchao/prototype/calibration_flow/static_quant.py

Reviewers:

Subscribers:

Tasks:

Tags:

### CODE REVIEW

**Comment by jerryzh168** [comment]
cc @drisspg this flow can be used for smoothquant as well, probably also float8

File: torchao/prototype/calibration_flow/static_quant.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
+"""
+Demo for static quantization flow
+"""
+import torch
+import copy
+
+from torch.ao.quantization.observer import MinMaxObserver, PerChannelMinMaxObserver
```
Comment (vkuzo) [inline torchao/prototype/calibration_flow/static_quant.py:7]: it's kind of weird to import the old quantization flows from the new ones, maybe add a TODO to fix this in the future

File: torchao/prototype/calibration_flow/static_quant.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    weight_obs = copy.deepcopy(weight_obs)
+    _replace_with_custom_fn_if_matches_filter(model, replacement_fn, _is_linear)
+
+def to_quantized_(model):
+    _is_calibrating_linear = lambda m, fqn: isinstance(m, CalibratingLinear)
+
+    replacement_fn = lambda m: QuantizedLinear.from_calibrating(m)
+    _replace_with_custom_fn_if_matches_filter(model, replacement_fn, _is_calibrating_linear)
+
+act_obs = MinMaxObserver(dtype=torch.uint8, qscheme=torch.per_tensor_affine).to("cuda")
```
Comment (vkuzo) [inline torchao/prototype/calibration_flow/static_quant.py:91]: i'd vote for rewriting this stuff without having to use old concepts like `qscheme`, instead of trying to reuse the code

File: torchao/prototype/calibration_flow/static_quant.py (hunk 1 above)
Comment (jerryzh168) [reply torchao/prototype/calibration_flow/static_quant.py:7]: yeah sounds good

File: torchao/prototype/calibration_flow/static_quant.py (hunk 2 above)
Comment (jerryzh168) [reply torchao/prototype/calibration_flow/static_quant.py:91]: agree, this is just an example, the end goal is to implement a generic observer for blockwise quantization

**Comment by jerryzh168** [comment]
@vkuzo @msaroufim I added a comment to update observers later, please take a look again

File: torchao/dtypes/affine_quantized_tensor.py [hunk 3]
```diff
@@ -228,6 +228,49 @@ def from_float(
             dtype=input_float.dtype
         )
 
+    @classmethod
+    def from_float_static(
+        cls,
+        input_float: torch.Tensor,
+        scale: torch.Tensor,
+        zero_point: torch.Tensor,
+        block_size: Tuple[int, ...],
```
Comment (msaroufim) [inline torchao/dtypes/affine_quantized_tensor.py:237]: what's the full type definition for this?

File: torchao/dtypes/affine_quantized_tensor.py [hunk 4]
```diff
@@ -228,6 +228,49 @@ def from_float(
...
+        target_dtype: torch.dtype,
+        quant_min: Optional[int] = None,
+        quant_max: Optional[int]  = None,
+        zero_point_domain: ZeroPointDomain = ZeroPointDomain.INT,
+        extended_layout: str = "plain",
```
Comment (msaroufim) [inline torchao/dtypes/affine_quantized_tensor.py:242]: this seems like it's using the old API? So I'm guessing you're landing this PR first?

File: torchao/dtypes/affine_quantized_tensor.py [hunk 5]
```diff
@@ -228,6 +228,49 @@ def from_float(
...
+        # TODO: this is only for "tensor_core_tiled", need to figure out
+        # the proper API for this arg
+        inner_k_tiles: Optional[int] = None,
+    ):
+        original_shape = input_float.shape
+        if extended_layout == "tensor_core_tiled":
+            orig_out_features, orig_in_features = input_float.shape
+            in_features = find_multiple(orig_in_features, 1024)
```
Comment (msaroufim) [inline torchao/dtypes/affine_quantized_tensor.py:250]: Re the 1024 and 8 padding heuristics as well this is what the NVIDIA docs say https://x.com/marksaroufim/status/1621580671776092160

So this heuristic tends to be dtype and device dependent - it's possible 1024 and 8 are fine but that would be right mostly out of luck

File: torchao/prototype/calibration_flow/static_quant.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
+"""
```
Comment (msaroufim) [inline torchao/prototype/calibration_flow/static_quant.py:1]: no test?

File: torchao/prototype/calibration_flow/static_quant.py (hunk 2 above)
Comment (msaroufim) [reply torchao/prototype/calibration_flow/static_quant.py:94]: What's the blocker for getting rid of the torch.ao dependency now? Are there toy observers we can include for now?

File: torchao/prototype/calibration_flow/static_quant.py [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+
+before_quant = m(*example_inputs)
+
+to_calibrating_(m, act_obs, weight_obs)
+# calibrating / training
+for _ in range(10):
+    m(*example_inputs)
+
+after_obs = m(*example_inputs)
+to_quantized_(m)
```
Comment (msaroufim) [inline torchao/prototype/calibration_flow/static_quant.py:105]: not sure I love the `to_calibrating` and `to_quantized` names or why not `calibrate()` and `quantize()`

**Review by msaroufim — CHANGES_REQUESTED** [review]

File: torchao/dtypes/affine_quantized_tensor.py (hunk 3 above)
Comment (jerryzh168) [reply torchao/dtypes/affine_quantized_tensor.py:237]: you mean `block_size`? this is just a tuple of variable length of ints

File: torchao/dtypes/affine_quantized_tensor.py (hunk 4 above)
Comment (jerryzh168) [reply torchao/dtypes/affine_quantized_tensor.py:242]: yeah, we can land the other PR first and then update this

File: torchao/prototype/calibration_flow/static_quant.py (hunk 2 above)
Comment (jerryzh168) [reply torchao/prototype/calibration_flow/static_quant.py:94]: I think we want to define a general observer that works for the new quant primitives in the end so that we can replace this.

we can define a toy observer I think. although I'm not sure why we want to get rid of torch.ao dep since we already depend on pytorch.

File: torchao/dtypes/affine_quantized_tensor.py (hunk 5 above)
Comment (jerryzh168) [reply torchao/dtypes/affine_quantized_tensor.py:250]: this should only be used by the tinygemm use case I think, so we should probably add some verifications

File: torchao/prototype/calibration_flow/static_quant.py (hunk 6 above)
Comment (jerryzh168) [reply torchao/prototype/calibration_flow/static_quant.py:1]: this is a demo for how static quantization can be implemented, instead of an API, should this be moved to some where else like `tutorials` folder?

File: torchao/prototype/calibration_flow/static_quant.py (hunk 7 above)
Comment (jerryzh168) [reply torchao/prototype/calibration_flow/static_quant.py:105]: `calibrate` is the name for the calibration process instead of the model transformation step I think, `quantize` is already used by the weight only and dynamic quant flow..

File: torchao/prototype/calibration_flow/static_quant.py (hunk 6 above)
Comment (msaroufim) [reply torchao/prototype/calibration_flow/static_quant.py:1]: Yeah perhaps docs is a better location for now

File: torchao/prototype/calibration_flow/static_quant.py (hunk 2 above)
Comment (msaroufim) [reply torchao/prototype/calibration_flow/static_quant.py:94]: It's more that torchao is supposed to be a full replacement. Id make an exception if some piece of code is very large and complex but that doesn't seem to be the case here

**Review by msaroufim — APPROVED** [review]
Approving assuming we move the static_quant.py file to either `docs` or `tutorials/` and do a fast follow to remove the `torch.ao` dependency

As far as renaming
* `to_calibrating` -> `insert_observers()`
* `to_quantized` -> `quantize_()`

File: torchao/quantization/quant_api.py [hunk 8]
```diff
@@ -259,12 +259,12 @@ def insert_subclass(lin):
 
     return insert_subclass
 
-def quantize_(model: torch.nn.Module, apply_tensor_subclass: Callable[[torch.Tensor], torch.Tensor], filter_fn: Optional[Callable[[torch.nn.Module, str], bool]]=None, set_inductor_config: bool=True):
+def quantize_(model: torch.nn.Module, apply_tensor_subclass: Callable[[torch.nn.Module], torch.nn.Module], filter_fn: Optional[Callable[[torch.nn.Module, str], bool]]=None, set_inductor_config: bool=True):
```
Comment (jerryzh168) [inline torchao/quantization/quant_api.py:262]: @msaroufim I made some changes to `apply_tensor_subclass` to accommodate static quant use cases, please take a look again
